### DISCUSSION TITLE
Aggregating Collections in Projections w. Quarkus/Hibernate

### DISCUSSION THREAD

**Discussion by ru4ert** [quarkusio/quarkus#39331] [Q&A] [answered]
Hello everyone,

I'm working on a project using Quarkus with Hibernate Panache, and I've encountered an issue related to entity projection, particularly when dealing with @OneToMany associations.

Here's a simplified version of my entity setup:

``` java
@Entity
class User {
  @OneToMany(mappedBy = "user")
  private List<ShoppingCartEntity> virtualShoppingCarts;
}
```
In this setup, a User entity is associated with multiple ShoppingCartEntity instances. For a specific user, let's say there are 3 associated virtualShoppingCarts.

To optimize my query performance, I'm using a projection to fetch user data:

``` java
@RegisterForReflection
public class UserProjection {
  public UserProjection(@ProjectedFieldName("virtualShoppingCarts") Object vsc) {}
}
```
And here's how I'm fetching the user:

``` java
var res = User.find("id.userId = :userId", Parameters.with("userId", userId))
                      .project(UserProjection.class)
                      .list();
```
Although I'm fetching a single user, the result (res) unexpectedly has a size of 3, presumably because of the 3 virtualShoppingCarts. However, my intention was to aggregate or collect these associated entities into a single projection instance.

Could someone guide me on how to correctly aggregate the @OneToMany collection in the projection when using Quarkus/Hibernate/Panache? Is there a specific approach or pattern to achieve this?

Any advice or examples would be greatly appreciated. Thank you in advance for your help!

[a comment by an automated account is not shown]

**Reply by yrodiere** [reply]
@FroMage @loicmathieu This seems to be related to Panache, so... when you have some time :)

**Reply by FroMage** [reply]
Projection does not change the result size, you get one projected result per result. What you're looking for is agreggation. The easiest is to construct the `UserProjection` type yourself from the `List<ShoppingCartEntity>` result. Performance will be exactly the same.

**Reply by ru4ert** [reply]
Can you provide an example/reference? I don't get what you mean. @FroMage

**Reply by FroMage** [reply]
```java
var res = new UserProjection(UserAreaJoin.find("id.userId = :userId", Parameters.with("userId", userId))
                      .list());
```

**Reply by ru4ert** [reply]
@FroMage but then i ran into  1 + n issues since i cannot define which fields i really need, right?

**Reply by ru4ert** [reply]
I want to use query projection that i only select couple fields and not the whole entity

**Reply by FroMage** [reply]
Ah, OK, but on what entity, though? `User` or `ShoppingCartEntity`? And what is the `UserAreaJoin` entity?

**Reply by ru4ert** [reply]
`ShoppingCartEntity` since its a `@OneToMany`.

I renamed the `UserAreaJoin` for simplification.

**Reply by ru4ert** [reply]
![image](https://github.com/quarkusio/quarkus/assets/47078678/72de1c0f-4596-494f-9768-1f4d5e579614)

**Reply by ru4ert** [reply]
If i have 6 vcs for a user and i make a select like this;
```
SELECT user.id, sc.id FROM User user
INNER JOIN ShoppingCartEntity sc ON sc.id = user.vsc
```

I get a ResultSet of this:
RS =>

```
USER ID                              | VSC Id
6dfa2a50-7be6-49cf-b95d-c324ec4fa7d3 | 9a1064f4-530c-4ab6-a10b-34fed5e9ba00
6dfa2a50-7be6-49cf-b95d-c324ec4fa7d3 | 001e84ad-b641-4aa8-bdf1-8fdb0a819ede
6dfa2a50-7be6-49cf-b95d-c324ec4fa7d3 | 768facaa-1985-414d-b830-552cdae6f04a
6dfa2a50-7be6-49cf-b95d-c324ec4fa7d3 | acc1e9b6-9ae4-4fc0-acac-d833eb5cec73
6dfa2a50-7be6-49cf-b95d-c324ec4fa7d3 | 2882ee38-36af-4f03-9197-1c5e0feec2ec
6dfa2a50-7be6-49cf-b95d-c324ec4fa7d3 | 4997e722-0160-454e-aa6f-35eb338755f0
```

For this reason is called 6 times.
`public UserProjection(@ProjectedFieldName("virtualShoppingCarts") Object vsc) {}`


Imaging the user has more `@OneToMany` joins. this is quite exponential.

The only solution is that i fetch `@OneToMany` afterwards. I want a less low level solution (if available)

**Reply by FroMage** [reply]
What about:

```java
var res = ShoppingCartEntity.find("userId", userId)
                      .project(UserProjection.class)
                      .list();
```

**Reply by FroMage** [reply]
I mean, if you want to limit the fields selected on `ShoppingCartEntity` then project on this class.

**Reply by ru4ert** [reply]
@FroMage 
I can do this. But if i have a entity with lets say 6 `@OneToMany` i need 7 calls to the db right?

1) Project the user
2) Project the 6 `@OneToMany`s

**Reply by FroMage** [reply]
6 different relations, you mean? Or 6 elements in a single `@OneToMany` relations?

**Reply by ru4ert** [reply]
@FroMage 6 relations with x elements in.

**Reply by FroMage** [reply]
Ah, OK. Well, as far as I know, there's no way to do recursive projections in Hibernate ORM.

So you can make a `User` projection to leave some fields out, but you can't turn `@OneToMany` fields into their own projections in the same projection.

Is that correct, @gavinking @yrodiere @DavideD ?

**Reply by yrodiere** [reply]
> Ah, OK. Well, as far as I know, there's no way to do recursive projections in Hibernate ORM.
> 
> So you can make a `User` projection to leave some fields out, but you can't turn `@OneToMany` fields into their own projections in the same projection.
> 
> Is that correct, @gavinking @yrodiere @DavideD ?

No idea.

@beikov?

**Reply by beikov** [reply, marked as answer]
Hibernate ORM only supports creating projections (through the use of the HQL constructor syntax) of flat results i.e. on projection object per result row. To map collection associations to into sub-projections, one will have to structure flat objects into hierarchical ones.

You can use [Blaze-Persistence Entity-Views](https://github.com/Blazebit/blaze-persistence?tab=readme-ov-file#entity-view-usage) for this purpose though, which does all of that for you and also allows to configure fetch strategies for such collection associations to fine tune the performance even further.

**Reply by ru4ert** [reply]
Ok i will have a look at this. In native queries i can use the fns provided by a db.

**Reply by FroMage** [reply]
Thanks @beikov !
